Initial version with Django and Vue (#2)
melexis/traceability-viewer: README.md

> # traceability-viewer

Reviewer: Please delete this top-level title and change the level of all others so that they are rendered larger.

Author: Resolved in bad373c6c3bf6e5eb1e3467b736dfc50a3da6f1e

## Table of Contents

- [Getting Started](#Getting-Started)
- [Configuration](#Configuration)
  - [traceability_export](#traceability_export)
  - [html_documentation_root](#html_documentation_root)
  - [layered](#layered)
  - [layers](#layers)
  - [item_colors](#item_colors)
  - [link_colors](#link_colors)
  - [backwards_relationships](#backwards_relationships)
  - [visualised_properties](#visualised_properties)
- [Usage](#Usage)

> 

Reviewer: Please add an "Overview" chapter that explains what the traceability viewer can be used for, what it does (it visualizes a collection of traceable items in a network graph), and that it uses the JSON database of our traceability plugin as input.

This should give the reader an idea whether this tool could be useful to them or not.

Author: Resolved in 388e0b3c04206b709422cca54a34ccaab5d2bd77

## Getting Started

First, you need to install docker compose: https://docs.docker.com/compose/install/linux/#install-the-plugin-manually

<!-- https://docs.docker.com/engine/reference/commandline/compose_up/ -->

After installation, copy the JSON file that contains the data to the project folder and copy the `.env.example` to `.env`.
```
# copy traceability_export.json file to project folder
cp /path/to/traceability_export.json .
# copy example .env to your .env
cp .env.example .env
```
Lastly, use docker compose to start the application.

```
docker compose up --build --remove-orphans

> ```

Reviewer: Please explain why we chose to make use of the Docker Compose plugin and that the `.env` file is meant to store and define environment variables and requires modification.

How can the web application and database be accessed after starting the application?

Author: Resolved in b676f408445bfd41ceb439da57a63ac4d72f77ac and 11f1ae6b7cbd81beaaadf9b001373c7e1cdd7ffa

## Configuration

The configuration takes the form of a YAML file. The [config.yml](config.yml) is an example that requires modification.

> ### traceability_export

Reviewer: This is not a configuration parameter. Please move this to section "getting started".

Author: Resolved in d6015a72219f992e6826a558aa3cedc14dd71da2

> [!IMPORTANT]
> `traceability_export` is a required configuration parameter.

This is the path to the JSON file that contains all the data. The items in the JSON file needs to have a name and targets.

It should be structured as (this is the minimum required):
```
[
    {
        "name" : "<name-of-item>",
        "targets": {
            "<relation-type-1>": [
                "<target-item-name-1>",
                "<target-item-name-2>",
                ...
            ],
            "<relation-type-2>": [
                "<target-item-name-3>",
                "<target-item-name-4>",
                ...
            ],
        }
    },
    ...
]
```

> When you use the [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/index.html), you can [export](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export) the documentation items as a JSON.

Reviewer: ```suggestion
Our Sphinx plugin [mlx.traceability](https://melexis.github.io/sphinx-traceability-extension/index.html) can be configured to
[export](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export) its collection of traceable documentation items in JSON format.
```

### html_documentation_root

> [!NOTE]

> > `html_documentation_root` is not a required configuration parameter.

Reviewer: ```suggestion
> `html_documentation_root` is an optional configuration parameter.
```
Please adjust this everywhere.

Author: Resolved in 892a28914e61675a6379ff69b3a4e8c11bd59fed

> The html root to the documentation of that item. This can also be the path to a local html folder.

Reviewer: ```suggestion
The root of the the HTML project documentation; either a pathname or a URL.
```

> This is only workable when you use the exported JSON of the [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export-to-json).

Reviewer: What do you mean by this sentance?

Author: I think it is only useful when you use the traceability plugin, but actually it can be any documentation that is composed of attributes of that element

Reviewer: The JSON database must contain the path to the RST document and the item ID/name must be appended to the HTML filename with a `#`. The traceability plugin does both.

```suggestion
This configuration parameter can only be used if [the JSON database](###traceability_export) contains a ``"document"`` attribute with the relative path to the reST file in which the item is defined,
and if the item ID/name is appended to the HTML filename with a `#`. 
Our [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export-to-json) does both.
```

Author: Suggestion applied in a7001e19f8f006e609f1a4d6a711a01590a42923

The suggestions only contain the names of nodes because there is nothing else allowed.

### Group Filter

The group filter are based on the configuration file. If `layers` is given, these unique groups are groups to filter on. When `layers` is nog defined, the groups are based on the groups of the `item_colors`.
The filter button gives a filtring on the specified group (the name of the button), with all their connecting nodes.
An example of such an filter is shown in the following image.

![filter](images/filter.png)

The legend is based on the visible nodes and links. When diffrent nodes or links are added, the legend will be updated.

### Interactions

#### Hide Groups

When clicking a group of the legend, this group will disappear. When clicking again, it reverts to previous.

![hide group](images/hide_group.png)

#### Hide Links

When clicking a link of the legend, this link will disappear. When clicking again, it reverts to previous.

![hide group](images/hide_link.png)

#### Selecting a Node

When clicking a node, this node will be selected and the info will be visible in the top left corner.

![select node](images/select_node.png)

> The image shows the info of the selected node. There are two more buttons showed as well. These two buttons pop up when a node is selected. These will be explained in [highlight connecting nodes](####highlight_connecting_nodes) and [Add/show connecting nodes](####add/show_connecting_nodes).

Reviewer: ```suggestion
You can click on a node to select it. Additional info about the node will be displayed in the top left corner and
two additional buttons pop up:

- [Highlight Connecting Nodes](####Highlight-Connected-Nodes)
- [Add/Show Connecting Nodes](####Add/Show-Connecting-Nodes)

![select node](images/select_node.png)
```
Spaces must be replaced by `-` when referencing titles.

#### Hovering a Node

When hovering a nodes, the name of that node is shown.
In the previous image, the black node at the bottom was hovered by the mouse.

#### HTML link in Info of Node

When `html_documentation_root` is specified, the name of the node will be a link to the documentation of that specific node.

#### Highlight Connected Nodes

> When the button with the eye is clicked, the nodes connecting to the selected node will be highlighted. The rest of the nodes get a lower opacity.

Reviewer: "nodes connecting to the selected node"

How far? Are only directly connected nodes highlighted?

```suggestion
The button with the eye-icon highlights all nodes that are directly connected to the selected node and makes other nodes transparent.
```

![highlight connected nodes](images/highlight_connected_nodes.png)

#### Add/Show Connecting Nodes

When the button **Add/show connecting nodes** is clicked, the nodes connecting to the selected node will be obtained. It will add new nodes or links, if there are any.

Before pushing the button:
![before adding](images/before_adding.png)

After pushing the button:
![add connecting nodes](images/add_connecting_nodes.png)

#### Node labels

The labels/names of the nodes are shown when the checkbox of **Nodes with labels** is active.

![show labels](images/show_labels.png)

#### Zoom buttons

There are three buttons for zooming, respectively zoom in, zoom out and zoom till the network fits the canvas.

![zoom buttons](images/zoom.png)

#### Errors

> When an error occurs you get a notification at the top of the page as shown in following example where an invalid Cypher query is submitted:

Reviewer: ```suggestion
When an error occurs, a notification pops up at the top of the page. In the example below, an invalid Cypher query was submitted:
```

Author: Suggestion applied in commit 9781b6a7c6706a1562ff853f01c4114f28604b40

![error](images/error.png)

You can show the error by clicking **Show error message**:

![error](images/show_error.png)

You can close the error by clicking **X**.

> If the error is not closed and another error occurs, all errors will be listed at the top of the page.
> You will be redirected to the hight of the newest error.

Reviewer: ```suggestion
If the error is not closed and another error occurs, the new error pops up at the top.
```

Author: Suggestion applied in commit 7ee06db64ba6d1a14607ea363ebe43ae66775260

## Bugs

> 1. When showing the labels/names of the nodes, there is allways one node that turns black.
> 1. When using the input fields, the autocompletion list will always look at the last part of the sentence, so when adding something in the middle the autocompletion will stay the same (focused on the last part).
> 
> 

Reviewer: ```suggestion
1. When showing the labels/names of the nodes, there is always one node that turns black.
2. When using the input fields, the autocompletion list will always look at the last part of the sentence. So, when adding something in the middle, the autocompletion does not adapt.
```

Author: Suggestion applied in commit 7ee06db64ba6d1a14607ea363ebe43ae66775260

> ## Features

Reviewer: ```suggestion
## Planned Features
```

Author: Suggestion applied in commit 065f4d44857d67b29d971145dcb146c2da52efd9

> - Errors could be popping up, you need to close it before you can go further. That way the user can't miss it.

Reviewer: Do you mean that certain errors should become blocking?

Author: I mean to use for example modals: https://getbootstrap.com/docs/5.3/components/modal/#static-backdrop
or toasts: https://getbootstrap.com/docs/5.3/components/toasts/#live-example

A modal that needs to be closed to go further of toasts that stay on your screen even when scrolling

Reviewer: I like Bootstrap's toasts. It doesn't interrupt the user's workflow compared to scrolling to the top or a modal that blocks progress.

- Faster rendering; open to suggestions
- Instead of disabling the button when submitted, stop the previous request when another submit is done.

> - A node that needs a layer but the group is not specified, better to place it on the same hight as the first node that is connected to it.

Reviewer: ```suggestion
- A node that needs a layer but the group is not specified, better to place it on the same height
  as the first node that is connected to it.
```
Please rephrase this sentence.

Author: Resolved in 3bc1af842099640e02bb7e825ab4b318196169b1

> - Reset button when in a filter.

Reviewer: ```suggestion
- Add a reset button to remove/undo the active filter.
```

Author: Suggestion applied in 81ae508d368c73c4aa3a89a1b19e08fe07823e90

> [!IMPORTANT]
> `backwards_relationships` is a required configuration parameter.

A dictionary where the key is the backwards relationship and the value the forward relationship. To reduct the amount of links, only the forward relationhsips (values) will be displayed.

### visualised_properties

> [!NOTE]
> `visualised_properties` is not a required configuration parameter.

The properies of the clicked node that need to be visualised in the as info about that node.

If the `visualised_properties` and [`html_documentation_root`](#html_documentation_root) are not defined , it automatically adds following code to the configuration file.

```
visualised_properties:
    - name
```

If [`html_documentation_root`](#html_documentation_root) is defined it automatically adds following code instead of the previous one.

```
visualised_properties:
    - name
    - document
```

If [`html_documentation_root`](#html_documentation_root) and `visualised_properties` are defined but `document` is not in the list, it automatically adds `document` to the end of the list.

If `visualised_properties` is defined but `name` is missing, it automatically adds `name` to the end of the list

> 

Reviewer: ```suggestion
> `visualised_properties` is an optional configuration parameter.

A list of all item properties to display after selecting a node. By default, its name/ID will be displayed, and,
in case [`html_documentation_root`](#html_documentation_root) has been configured,
the link to the item definition in your project documentation,.
```

Author: Suggestion applied in commit 815d068e34420389f6cb96f9272f34c59d13ea1d

# Table of Contents

- [Overview](#overview)
- [Getting Started](#getting-started)
  - [JSON Database](#json-database)
- [Configuration](#configuration)
  - [html_documentation_root](#html_documentation_root)
  - [layered](#layered)
  - [layers](#layers)
  - [item_colors](#item_colors)
  - [link_colors](#link_colors)
  - [backwards_relationships](#backwards_relationships)
  - [visualised_properties](#visualised_properties)
- [Usage](#usage)
  - [Home Page](#home-page)
    - [Cypher Query](#cypher-query)
    - [Search](#search)
  - [Group Filter](#group-filter)
  - [Interactions](#interactions)
    - [Hide Groups](#hide-groups)
    - [Hide Links](#hide-links)
    - [Selecting a Node](#selecting-a-node)
    - [Hovering a Node](#hovering-a-node)
    - [HTML link in Info of Node](#html-link-in-info-of-node)
    - [Highlight Connected Nodes](#highlight-connected-nodes)
    - [Add/Show Connecting Nodes](#addshow-connecting-nodes)
    - [Node labels](#node-labels)
    - [Zoom buttons](#zoom-buttons)
    - [Errors](#errors)
- [Bugs](#bugs)
- [Features](#features)

# Overview

> The traceability visualizes a collection of traceable items in a network graph. This allows the user to gain a better understanding of the traceability between the different items. The network can be layered or force-directed. This and more options are [configurable](#configuration).
> 
> It's originally designed to use the JSON database of our traceability plugin as input but it can also be used with other JSON databases as well. The minimum required JSON database is given in [this section](#json-database).

Reviewer: ```suggestion
This traceability-viewer, a standalone tool, visualizes a collection of traceable documentation items in a network graph.
This allows the user to gain a better understanding of the relationships between different items.
Each node in the network graph represents an item, and can be grouped in layers, based on its identifier,
or force-directed. This and more options are [configurable](#configuration).

It was originally designed to use the JSON database exported by our
[Sphinx traceability plugin](https://github.com/melexis/sphinx-traceability-extension) as input
but any JSON database can work if it meets the minimum requirements, explained in [this section](#json-database).
```

Before pushing the button:
![before adding](images/before_adding.png)

After pushing the button:
![add connecting nodes](images/add_connecting_nodes.png)

### Node labels

The labels/names of the nodes are shown when the checkbox of **Nodes with labels** is active.

![show labels](images/show_labels.png)

### Zoom buttons

There are three buttons for zooming, respectively zoom in, zoom out and zoom till the network fits the canvas.

![zoom buttons](images/zoom.png)

### Errors

When an error occurs, a notification pops up at the top of the page. In the example below, an invalid Cypher query was submitted:

![error](images/error.png)

You can show the error by clicking **Show error message**:

![error](images/show_error.png)

You can close the error by clicking **X**.
If the error is not closed and another error occurs, all errors will be listed at the top of the page.

# Bugs

1. When showing the labels/names of the nodes, there is always one node that turns black.
1. When using the input fields, the autocompletion does not add the selected word at the right place, when adding something in the middle.

> 1. When data is obtained via a Cypher query, some links are not visible. When clicking the eye-icon, these nodes that are connected are visible but the links aren't. You can solve it by clicking on the `Add/show connecting nodes` button.

Reviewer: ```suggestion
1. When data is obtained via a Cypher query, some links are not visible. When clicking on the eye-icon, connected
   nodes are visible but the links are not. You can solve it by clicking on the `Add/show connecting nodes` button.
```
Improved readability and more focus on **not** visible.

# Overview

This traceability-viewer, a standalone tool, visualizes a collection of traceable documentation items in a network graph.
This allows the user to gain a better understanding of the relationships between different items.
Each node in the network graph represents an item, and can be grouped in layers, based on its identifier,
or force-directed. This and more options are [configurable](#configuration).

It was originally designed to use the JSON database exported by our
[Sphinx traceability plugin](https://github.com/melexis/sphinx-traceability-extension) as input
but any JSON database can work if it meets the minimum requirements, explained in [this section](#json-database).

# Getting Started

We use the tool Docker Compose to define and share multi-container applications. This tool makes use of a `docker-compose.yml` file. The `.env` file is meant to store and define environment variables, which are used in the `docker-compose.yml` file and in Python scripts.

First, you need to [install Docker Compose](https://docs.docker.com/compose/install/linux/#install-the-plugin-manually).

<!-- https://docs.docker.com/engine/reference/commandline/compose_up/ -->

> After installation, copy the JSON file that contains the data to the project folder and copy the `.env.example` to `.env`. These environment variables requires modification.

Reviewer: ```suggestion
After installation, copy the JSON database file to this project's folder, copy the `.env.example` to `.env` and customize it.
```

![hide group](images/hide_group.png)

### Hide Links

When clicking a link of the legend, this link will disappear. When clicking again, it reverts to previous.

![hide group](images/hide_link.png)

### Selecting a Node

You can click on a node to select it. Additional info about the node will be displayed in the top left corner and
two additional buttons pop up:

- [Highlight Connecting Nodes](#highlight-connected-nodes)
- [Add/Show Connecting Nodes](#addshow-connecting-nodes)

![select node](images/select_node.png)

### Hovering a Node

When hovering a nodes, the name of that node is shown.
In the previous image, the black node at the bottom was hovered by the mouse.

### HTML link in Info of Node

When `html_documentation_root` is specified, the name of the node will be a link to the documentation of that specific node.

### Highlight Connected Nodes

The button with the eye-icon highlights all nodes that are directly connected to the selected node and makes other nodes transparent.

![highlight connected nodes](images/highlight_connected_nodes.png)

### Add/Show Connecting Nodes

> When the button **Add/show connecting nodes** is clicked, the nodes connecting to the selected node will be obtained. It will add new nodes or links, if there are any.

Reviewer: ```suggestion
When you selected a node and click the button **Add/show connecting nodes**, all nodes that are connected to it are obtained and displayed, if there are any.
```
Let's avoid the passive tense in cases like this. It improves readability.

### Node labels

The labels/names of the nodes are shown when the checkbox of **Nodes with labels** is active.

![show labels](images/show_labels.png)

### Zoom buttons

There are three buttons for zooming, respectively zoom in, zoom out and zoom till the network fits the canvas.

![zoom buttons](images/zoom.png)

### Errors

When an error occurs, a notification pops up at the top of the page. In the example below, an invalid Cypher query was submitted:

![error](images/error.png)

You can show the error by clicking **Show error message**:

![error](images/show_error.png)

You can close the error by clicking **X**.
If the error is not closed and another error occurs, all errors will be listed at the top of the page.

# Bugs

1. When showing the labels/names of the nodes, there is always one node that turns black.
1. When using the input fields, the autocompletion does not add the selected word at the right place,
when adding something in the middle.
1. When data is obtained via a Cypher query, some links are not visible.
When clicking on the eye-icon, connected nodes are visible but the links are not.
You can solve it by clicking on the `Add/show connecting nodes` button.

> 1. Hiding a group after selecting a node from that same group, the info of that node is still there.

Reviewer: ```suggestion
1. When hiding a group while a node of that group was selected, the info of that node remains visible.
```

1. When showing the labels/names of the nodes, there is always one node that turns black.
1. When using the input fields, the autocompletion does not add the selected word at the right place, when adding something in the middle.
1. When data is obtained via a Cypher query, some links are not visible. When clicking on the eye-icon, connected nodes are visible but the links are not. You can solve it by clicking on the `Add/show connecting nodes` button.

> 

Reviewer: GitHub rendered it just fine before this change of yours. Why did you change it? Long line lengths are not very nice to review and maintain.

https://github.com/melexis/traceability-viewer/blob/620324f5ca98fd9f79335a618733a09e96083e7e/README.md

![image](https://github.com/melexis/traceability-viewer/assets/28319872/539923d8-2764-4500-8f20-26992019932d)

Author: I removed the newline in the middle of the sentence because it would result in this:
![image](https://github.com/melexis/traceability-viewer/assets/62558419/61ac8469-8344-410f-a927-7e92fc40747d)

But I can add newlines after a sentence?

Author: I think I resolved it in 30a2aedf809fbcfa11292f6d77f54dfaec8458c7

Reviewer: This syntax seems to work in Markdown. FYI In reStructuredText, you would need to align the text:
```
- Bullets are "-", "*" or "+".
  Continuing text must be aligned
  after the bullet and whitespace
  ```
  https://docutils.sourceforge.io/docs/user/rst/quickref.html#bullet-lists
